### PR TITLE
add type annotations to torch.nn.modules.module

### LINKED ISSUE #49044: Enable torch.nn.modules.module typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.nn.modules.module]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #49044

### CODE REVIEW

**Comment by rgommers** [comment]
@guilhermeleobas could you also add comments here about the ignores you added? That helps both with review (reproducing the failures and using `reveal_type` is my alternative), and when the next person has to touch that bit of code.

**Comment by rgommers** [comment]
Also, in your next commit message, can you add that it closes gh-48640?

**Comment by guilhermeleobas** [comment]
Hi @rgommers, thanks for the review. I've addressed your comments

**Review by rgommers — APPROVED** [review]
Looks good now, thanks @guilhermeleobas.

CI notifications seem stuck, but CI is done - see https://app.circleci.com/pipelines/github/pytorch/pytorch?branch=pull%2F49045. Failures are unrelated.

File: torch/fx/symbolic_trace.py
Comment (walterddr) [inline torch/fx/symbolic_trace.py:93]: why cant we just use different variable names?

File: torch/nn/modules/module.py
Comment (walterddr) [inline torch/nn/modules/module.py:1205]: I am not python expert. but cant we use `del load` to remove the function reference itself? (after all the load has been done, yes?)

Comment (guilhermeleobas) [reply torch/nn/modules/module.py:1205]: Yes!

File: torch/fx/symbolic_trace.py
Comment (guilhermeleobas) [reply torch/fx/symbolic_trace.py:93]: Yes, that works too

**Comment by guilhermeleobas** [comment]
Do not merge this PR until one checks if the annotations introduce any regression. See:
https://github.com/pytorch/pytorch/pull/49564#issuecomment-752916072

File: torch/distributed/nn/api/remote_module.py
Comment (walterddr) [inline torch/distributed/nn/api/remote_module.py:280]: shouldn't we remove this type ignore as well since we fixed the return Parameter

**Review by walterddr — COMMENTED** [review]
picky back on this PR - i dont think there's any additional item that needs to be test with JIT-ability for this change. am I missing any? @malfet

File: torch/distributed/nn/api/remote_module.py
Comment (guilhermeleobas) [reply torch/distributed/nn/api/remote_module.py:280]:
```
torch/distributed/nn/api/remote_module.py:280: error: Missing return statement  [return]
```

Comment (walterddr) [reply torch/distributed/nn/api/remote_module.py:280]: lol. Got it.

**Review by walterddr — APPROVED** [review]
looks good. I dont see any additional JIT test required.

**Comment by walterddr** [comment]
actually please rebase master --> seems like merge conflict is preventing fb internal CI to run

**Pull request closed without merging**
